### PR TITLE
When handling caret movement, fetch the new caret info after checking for a pending caret event.

### PR DESCRIPTION
**Pull request by jcsteh**
### Link to issue number:
None.

### Summary of the issue:
In Firefox's in-progress UIA implementation, moving with the cursor keys in editable text boxes sometimes reports the character at the old position of the caret instead of the new position of the caret. For example, on the line:
abcdef
with the cursor positioned on "a", pressing right arrow several times might read a, b, c instead of b, c, d.
I don't anticipate that NVDA will ever move to using UIA in Firefox. Nevertheless, this is an NVDA bug which might impact other things.

### Description of user facing changes
When navigating with the cursor keys in text boxes in applications which use UI Automation, NVDA no longer sometimes reports the wrong character, word, etc.

### Description of development approach
Previously, `EditableText._hasCaretMoved` fetched the caret before checking for a pending caret event. It then returned this fetched caret. The following race condition could occur:

1. In NVDA's main thread, NVDA queries the caret. The caret hasn't updated yet in the app, so this is the old position.
2. The app fires a caret event.
3. NVDA receives this caret event on a background UIA thread and queues it.
4. Back in NVDA's main thread, NVDA checks for a pending caret event. That check passes because of 3).
5. NVDA returns the old caret info retrieved in 1).

This can't happen with other APIs because they all queue events on the main thread, which can't be preempted during `_hasCaretMoved`. However, UIA queues events from a background thread.

### Testing strategy:
I tested caret navigation in various apps: normal Firefox, Edge, Windows Run dialog, Notepad++, Windows Settings, Windows Start Menu. Everything worked as expected.
I tested caret navigation in Firefox's in-progress UIA implementation. After this change, everything worked as expected.

### Known issues with pull request:
None.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved caret movement detection to prevent NVDA from reporting incorrect characters or words when navigating with cursor keys in text boxes using UI Automation.

### CODE REVIEW

File: source/editableText.py [hunk 1]
```diff
@@ -85,27 +85,34 @@ def _hasCaretMoved(self, bookmark, retryInterval=0.01, timeout=None, origWord=No
...
+				try:
+					newInfo = self.makeTextInfo(textInfos.POSITION_CARET)
+				except (RuntimeError, NotImplementedError):
+					newInfo = None
+				return (True, newInfo)
 			# If the focus changes after this point, fetching the caret may fail,
 			# but we still want to stay in this loop.
 			try:
 				newInfo = self.makeTextInfo(textInfos.POSITION_CARET)
 			except (RuntimeError,NotImplementedError):
```
Comment (seanbudd) [inline source/editableText.py:114]:
```suggestion
			except (RuntimeError, NotImplementedError):
```

**Review by seanbudd — APPROVED** [review]

**Comment by LeonarddeR** [comment]
I wonder whether we should re-investigate all cases where `caretMovementDetectionUsesEvents` is currently set to `False`, especially in `InaccurateTextChangeEventEmittingEditableText`. It could be that the assumption in there is/was just wrong.

**Comment by jcsteh** [comment]
It turns out that I can reproduce this in text boxes in web content in Edge when NVDA is using UIA for Edge.

**Comment by codeofdusk** [comment]
In particular, I'd like to play with this branch in UIA terminals. I disabled caret events there because when quickly backspacing, the last character of the prompt was sometimes eroneously read.

**Comment by seanbudd** [comment]
@jcsteh - does this mean this PR should not be merged and a different approach is found?

**Comment by jcsteh** [comment]
Sorry for the lack of clarity. I meant that the original issue (without this PR) can be reproduced in Edge + UIA. This PR should fix the problem for both Firefox and Edge when using UIA. I thought this worth noting because while UIA in Firefox is not likely to be used "in the real world", NVDA does have explicit support for UIA in Edge and that is the only option for certain use cases, so that increases the user impact of this bug from theoretical to practical.

**Comment by codeofdusk** [comment]
@burakyuksek Notepad is a UIA app.

**Comment by burakyuksek** [comment]
@codeofdusk I'm aware, however the issue still occurs.

**Comment by jcsteh** [comment]
This PR should apply to all UIA implementations, so if this is still happening in Notepad for you, that would suggest it is a different issue. A couple of possibilities:

1. The cursor is taking too long to move, so NVDA times out and reports the old caret position.
2. I think there might be some bugs when moving the caret across wrapped lines. Are the issues you're seeing at the end of wrapped lines?

Otherwise, I have no idea.

**Pull request merged**